### ISSUE TITLE
[discussion] Trying to harmonize checks in FreeRDP

### ISSUE THREAD

**Issue by hardening** [FreeRDP/FreeRDP#4485]
Hi guys,
I'm opening this issue so that we try to find a common position on how to handle checks in FreeRDP. This is an attempt to deal with PR #4322 or #4240, where @akallabeth and I, we have different views on the subject.

So basically we have that kind of code

```C
static BOOL treat_case1(Type1 *arg1, Type2 *arg2) 
{
   ....
}

BOOL do_treatment(Type1 *arg1)
{
  Type2 t2;

  switch(arg1->id) {
    ....
    case 1:
      return treat_case1(t1, &t2);
    ....
  }

}
```

In the cited PRs @akallabeth added some checks in `treat_case1` to check for `arg1` and `arg2`. So that the function becomes:
```C
static BOOL treat_case1(Type1 *arg1, Type2 *arg2) 
{
    if (!arg1 || !arg2)
        return FALSE;

   .....
}
```

In my opinion these are totally useless and even counter-productive:

1. as you can easily see in this function, checking for `arg1` is not needed as we would have crashed before in `do_treatment()`, and checking for `arg2` is not needed either as it is a reference on a stack variable (and so have no chance to be NULL). So in theory none of these checks are needed;
1. for sure this will end up with consuming a lot of CPU if these kind of checks are generalized;
1. @akallabeth pointed some security improvements but you can have a security gain only when there's some place for it, and as stated in point 1 it's not the case here;
1. for someone reading the code, it looks like having `arg1` or `arg2` set to NULL is a legitimate case, while we clearly see that if we end up in this situation, we have a massive corruption that must not be handled by just returning `FALSE`, but more by exiting the program ASAP (`abort`);
1. adding the checks makes the code less clean, and gives the impression of self protecting code. A well designed code checks inputs in the functions that are public. Other functions just pre-require that the checks have been done, otherwise it's code protecting against itself. For me these checks are a mark of bad design, where the public boundary of logical block is not clear;
1. we all know that these unneeded checks have been added when tracking a segfault during a debugging sessions, and that the real bug has been found after (and so the checks have not caught anything);
1. if we wanna add some sanity checks anyway, these should be asserts (`assert(arg1 && arg2);`). From what I understood asserts have the good idea to expand to void when compiled in release mode;

I gave you my position, I guess some of you will not totally agree, anyway my point was more to find a common position and then do the appropriate changes to have a consistent code base: add extra checks all over the place à la @akallabeth, or operate a cleanup on self-protecting code à la @hardening.

Looking forward for your comments.

**Comment by hardening** [comment]
So my answers:

1. For sure compilers have made progress on optimization, but I don't see why we should add extra checks that would be _possibly_ removed at the end by the compiler. Especially when it makes the code less clean;
2. Well on most PR that I've commented on, it's internal functions of channel implementations, so clearly not API. Checking everything, everywhere is just the proof of a lack of control, and of unclear boundaries between public and private. All the comments that I have done are clearly checks for situations that can't happen.
3. I used `legitimate` as `possible value`. In my example `arg2 == NULL` is not a runtime error with an invalid user/remote host input, it's a design error. You're right there's a claim (from debian guys IIRC) that should be discussed, that there should not have an `exit` or `abort`  in a library. Anyway if we come to the point where we figure that arg2 is NULL what shall we do ? As we've seen it's not theoretically possible, so if we hit that case that mean that we're in a very tricky situation with probably a massive corruption of the stack, so I don't think it's the best idea to `return FALSE` just like if we weren't aware of the situation. I clearly think that we're taking some big risks by returning on a trashed stack (that would also make debugging more difficult, as you will jump back to a random place), and for sure your logging won't work at the function return. So the 'no exit in a library' rule is really questionable.
4. Well I'd say exactly the opposite, it's really fine that `assert` are gone in `release`. Debugging is to be done with debug versions;
